README.md
### 简介

flutter_tencentplayer_plus  是基于腾讯云点播封装的flutter版的播放器插件 提供video_player 相似的api, 是flutter_tencentplayer的优化版本 ,flutter_tencentplayer 开源项目是我和[大帅](https://github.com/qq326646683/flutter_tencentplayer)一起合作的一个开源项目,
 

### 支持功能
1. 支持直播源
2. 视频跳转
3. 切换视频源
4. 边下边播放
5. 清晰度切换
6. 设置播放速度



### 效果图
![](readme/android.gif)



### 1.Setup
    
  
    
    flutter_tencentplayer_plus: ${last_version}
    
    or
    
    flutter_tencentplayer_plus:
        git:
          url: https://github.com/yxwandroid/flutter_tencentplayer_plus.git
  
> For Android

1. project/android/build.gradle 添加依赖的aar：

```
def flutterProjectRoot = rootProject.projectDir.parentFile.toPath()
def plugins = new Properties()
def pluginsFile = new File(flutterProjectRoot.toFile(), '.flutter-plugins')
if (pluginsFile.exists()) {
    pluginsFile.withReader('UTF-8') { reader -> plugins.load(reader) }
}

allprojects {
    repositories {
        google()
        jcenter()
        flatDir {
            dirs "${plugins.get("flutter_tencentplayer")}android/libs"
        }
    }
}
```

2. AndroidManifest.xml 声明权限:

    ```
    <!--网络权限-->
    
    <uses-permission android:name="android.permission.INTERNET" />
    
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    
    <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
    
    <!--点播播放器悬浮窗权限-->
    
    <uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW" />
    
    <!--存储-->
    
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
    
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
    ```

> For Ios  



``` 

//项目的info.plist文件上添加如下权限 
<key>NSAppTransportSecurity</key>
	<dict>
		<key>NSAllowsArbitraryLoads</key>
		<true/>
  </dict>
```

### 2.Usage(TencentPlayer)


#### 1.初始化播放器

提供 asset、network、filePath、fileId四种方式播放


```dart
TencentPlayerController _controller;

_MyAppState() {
    listener = () {
      if (!mounted) {
        return;
      }
      setState(() {});
    };
}

initState() {
    _controller = TencentPlayerController.network('http://file.jinxianyun.com/testhaha.mp4', playerConfig: PlayerConfig())
    //_controller = TencentPlayerController.asset('static/tencent1.mp4')
    //_controller = TencentPlayerController.file('/storage/emulated/0/test.mp4')
    //_controller = TencentPlayerController.network(null, playerConfig: {auth: {"appId": 1252463788, "fileId": '4564972819220421305'}})
    ..initialize().then((_) {
        setState(() {});
      });
    _controller.addListener(listener);
}
```

#### 2.<font color=#0000FF >PlayerConfig</font> (播放器配置参数 )

```dart
_controller = TencentPlayerController.network(url, playerConfig: PlayerConfig())

```

Prop | Type | Default | Note
---|---|---|---
autoPlay | bool | true | 是否自动播放
loop | bool | false | 是否循环播放
headers | Map<String, String> |  | 请求头
cachePath | String |  | 缓存路径(边播放边下载)
progressInterval | int | 200 | 播放进度回调频率(毫秒)
startTime | int | 0 | 哪里开始播放(秒)
auth | Map<String, dynamic> |  | 云点播视频源appId&fileId



#### 3.<font color=#0000FF >TencentPlayerValue</font> (播放器回调)
```dart
Text("总时长：" + _controller.value.duration.toString())
```
Prop | Type | Note
---|---|---
initialized | bool | 是否初始化完成从而显示播放器
aspectRatio | double | 用来控制播放器宽高比
duration | Duration | 时长
position | Duration | 播放进度
playable | Duration | 缓冲进度
isPlaying | bool | 是否在播放
size | Size | 视频宽高
isLoading | bool | 是否在加载
netSpeed | int | 视频播放网速
rate | double | 播放速度
bitrateIndex | int | 视频清晰度

#### 4.<font color=#0000FF >Event</font> (播放器事件)

a.跳转进度

```dart
_controller.seekTo(Duration(seconds: 5));

```

b.设置播放速度

```dart
_controller.setRate(1.5); // 1.0 ~ 2.0

```

c.切换播放源

```dart
controller?.removeListener(listener);
controller?.pause();
controller = TencentPlayerController.network(url, playerConfig: PlayerConfig(startTime: startTime ?? controller.value.position.inSeconds));
controller?.initialize().then((_) {
  if (mounted) setState(() {});
});
controller?.addListener(listener);
```

d.切换清晰度(实质就是切换播放源)


# 3.Usage(Download)

> 离线下载, 支持断点续传(这里只支持m3u8视频、fileId), 支持多文件同时下载

#### 1.初始化下载器

```dart
DownloadController _downloadController;

_MyAppState() {
    downloadListener = () {
      if (!mounted) {
        return;
      }
      setState(() {});
    };
}

initState() {
    _downloadController = DownloadController('/storage/emulated/0/tencentdownload', appId: 1252463788);
    _downloadController.addListener(downloadListener);
}
```
#### 2.<font color=#0000FF >Event</font> (下载事件)

a. 下载

```dart
_downloadController.dowload("4564972819220421305", quanlity: 2);
// _downloadController.dowload("http://1253131631.vod2.myqcloud.com/26f327f9vodgzp1253131631/f4bdff799031868222924043041/playlist.m3u8");
```

b. 暂停下载

```dart
_downloadController.pauseDownload("4564972819220421305");
// _downloadController.stopDownload("http://1253131631.vod2.myqcloud.com/26f327f9vodgzp1253131631/f4bdff799031868222924043041/playlist.m3u8");

```

b. 取消下载

```dart
_downloadController.cancelDownload("4564972819220421305");
// _downloadController.cancelDownload("http://1253131631.vod2.myqcloud.com/26f327f9vodgzp1253131631/f4bdff799031868222924043041/playlist.m3u8");

```


#### 3.<font color=#0000FF >DownloadValue</font> (下载信息回调)
>因为支持多文件同时下载,回调以Map<String, DownloadValue>返回,key为url/fileId

Prop | Type | Note
---|---|---
downloadStatus | String | "start"、"progress"、"stop"、"complete"、"error"
quanlity | int | 1: "FLU"、2: "SD"、3: "HD"、4: "FHD"、5: "2K"、6: "4K"
duration | int |
size | int | 文件大小
downloadSize | int | 已下载大小
progress | int | 已下载大小
playPath | String | 下载文件的绝对路径
isStop | bool | 是否暂停下载
url | String | 下载的视频链接
fileId | String | 下载的视频FileId
error | String | 下载的错误信息



### 4.[Example](https://github.com/qq326646683/flutter_tencentplayer/blob/master/example/lib/main.dart)



### 5.更新内容 

     1,升级android,IOS  LiteAVSDK 6.8.7969
 
### 6.参考

[腾讯云点播android版本](https://github.com/tencentyun/SuperPlayer_Android/wiki)
[腾讯云点播ios版本](https://github.com/tencentyun/SuperPlayer_iOS)
[flutter_tencentplayer](https://github.com/qq326646683/flutter_tencentplayer
)
    









CHANGELOG.md
## 0.1.0

* 格式化代码


## 0.0.1

* 初始化项目


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:flutter_tencentplayer_plus/flutter_tencentplayer_plus.dart';
import 'package:flutter_tencentplayer_plus_example/home_page.dart';

//void main() => runApp(MyApp());
void main() => runApp(launch);

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  TencentPlayerController _controller;
  VoidCallback listener;

  DownloadController _downloadController;
  VoidCallback downloadListener;

  String videoUrl =
      'http://5815.liveplay.myqcloud.com/live/5815_89aad37e06ff11e892905cb9018cf0d4.flv';
  String videoUrlB =
      'http://5815.liveplay.myqcloud.com/live/5815_89aad37e06ff11e892905cb9018cf0d4_550.flv';
  String videoUrlG =
      'http://5815.liveplay.myqcloud.com/live/5815_89aad37e06ff11e892905cb9018cf0d4_900.flv';
  String videoUrlAAA = 'http://file.jinxianyun.com/2018-06-12_16_58_22.mp4';
  String videoUrlBBB = 'http://file.jinxianyun.com/testhaha.mp4';
  String mu =
      'http://devimages.apple.com.edgekey.net/streaming/examples/bipbop_4x3/gear2/prog_index.m3u8';
  String spe1 =
      'http://1252463788.vod2.myqcloud.com/95576ef5vodtransgzp1252463788/e1ab85305285890781763144364/v.f10.mp4';
  String spe2 =
      'http://1252463788.vod2.myqcloud.com/95576ef5vodtransgzp1252463788/e1ab85305285890781763144364/v.f20.mp4';
  String spe3 =
      'http://1252463788.vod2.myqcloud.com/95576ef5vodtransgzp1252463788/e1ab85305285890781763144364/v.f30.mp4';

  String testDownload =
      'http://1253131631.vod2.myqcloud.com/26f327f9vodgzp1253131631/f4bdff799031868222924043041/playlist.m3u8';
  String downloadRes =
      '/storage/emulated/0/tencentdownload/txdownload/2c58873a5b9916f9fef5103c74f0ce5e.m3u8.sqlite';
  String downloadRes2 =
      '/storage/emulated/0/tencentdownload/txdownload/cf3e281653e562303c8c2b14729ba7f5.m3u8.sqlite';

  @override
  void initState() {
    super.initState();

    addListener();
    initPlatformState();
  }

  addListener() {
//    listener = () {
//      if (!mounted) {
//        return;
//      }
//      setState(() {});
//    };
    downloadListener = () {
      if (!mounted) {
        return;
      }
      setState(() {});
    };
  }

  Future<void> initPlatformState() async {
    //点播
//     _controller = TencentPlayerController.network(null, playerConfig:  PlayerConfig(
//        auth: {"appId": , "fileId": ''}
//    ))
    _controller = TencentPlayerController.network(spe3,
        playerConfig: PlayerConfig(autoPlay: false))

//        _controller = TencentPlayerController.asset('static/tencent1.mp4')
//        _controller = TencentPlayerController.file('/storage/emulated/0/test.mp4')
      ..initialize().then((_) {
        setState(() {});
      });

//    _controller.addListener(listener);
    /// 下载目录  ios 的和android的不一致 注意在ios项目运行的时候传递ios的目录
    _downloadController = DownloadController(
        '/storage/emulated/0/tencentdownload',
        appId: 1252463788);
    _downloadController.addListener(downloadListener);
  }

  @override
  void dispose() {
    super.dispose();
//    _controller.removeListener(listener);
    _downloadController.removeListener(downloadListener);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Video Demo',
      home: ValueListenableBuilder(
        valueListenable: _controller,
        builder:
            (BuildContext context, TencentPlayerValue value, Widget child) {
          return Scaffold(
            body: Container(
              child: Column(
                children: <Widget>[
                  Container(
                    child: Stack(
                      alignment: AlignmentDirectional.center,
                      children: <Widget>[
//                        _controller.value.initialized ?
                        AspectRatio(
                          aspectRatio: _controller.value.aspectRatio,
                          child: TencentPlayer(_controller),
                        ),
                        Center(
                          child: _controller.value.isLoading
                              ? CircularProgressIndicator()
                              : SizedBox(),
                        ),
                      ],
                    ),
                  ),
                  Expanded(
                    child: ListView(
                      children: <Widget>[
                        Text(
                          "播放网速：" + _controller.value.netSpeed.toString(),
                          style: TextStyle(color: Colors.pink),
                        ),
                        Text(
                          "错误：" + _controller.value.errorDescription.toString(),
                          style: TextStyle(color: Colors.pink),
                        ),
                        Text(
                          "播放进度：" + _controller.value.position.toString(),
                          style: TextStyle(color: Colors.pink),
                        ),
                        Text(
                          "缓冲进度：" + _controller.value.playable.toString(),
                          style: TextStyle(color: Colors.pink),
                        ),
                        Text(
                          "总时长：" + _controller.value.duration.toString(),
                          style: TextStyle(color: Colors.pink),
                        ),
                        FlatButton(
                            onPressed: () {
                              _controller.seekTo(Duration(seconds: 5));
                            },
                            child: Text(
                              'seekTo 00:00:05',
                              style: TextStyle(color: Colors.blue),
                            )),
                        Row(
                          children: <Widget>[
                            FlatButton(
                                onPressed: () {
                                  _controller.setRate(1.0);
                                },
                                child: Text(
                                  'setRate 1.0',
                                  style: TextStyle(
                                      color: _controller.value.rate == 1.0
                                          ? Colors.red
                                          : Colors.blue),
                                )),
                            FlatButton(
                                onPressed: () {
                                  _controller.setRate(1.5);
                                },
                                child: Text(
                                  'setRate 1.5',
                                  style: TextStyle(
                                      color: _controller.value.rate == 1.5
                                          ? Colors.red
                                          : Colors.blue),
                                )),
                            FlatButton(
                                onPressed: () {
                                  _controller.setRate(2.0);
                                },
                                child: Text(
                                  'setRate 2.0',
                                  style: TextStyle(
                                      color: _controller.value.rate == 2.0
                                          ? Colors.red
                                          : Colors.blue),
                                )),
                          ],
                        ),
                        Row(
                          children: <Widget>[
                            FlatButton(
                                onPressed: () {
                                  _controller =
                                      TencentPlayerController.network(mu);
                                  _controller.initialize().then((_) {
                                    setState(() {});
                                  });
                                  //  _controller.addListener(listener);
                                },
                                child: Text(
                                  'm3u8点播',
                                  style: TextStyle(
                                      color:
                                          _controller.dataSource == videoUrlAAA
                                              ? Colors.red
                                              : Colors.blue),
                                )),
                            FlatButton(
                              onPressed: () {
                                _controller =
                                    TencentPlayerController.network(spe1);
                                _controller.initialize().then((_) {
                                  setState(() {});
                                });
                                // _controller.addListener(listener);
                              },
                              child: Text(
                                '普通点播',
                                style: TextStyle(
                                    color: _controller.dataSource == videoUrlBBB
                                        ? Colors.red
                                        : Colors.blue),
                              ),
                            ),
                          ],
                        ),
                        Row(
                          children: <Widget>[
                            Container(
                              margin: EdgeInsets.only(left: 15),
                              child: Text(
                                'm3u8点播 : ',
                                style: TextStyle(color: Colors.orange),
                              ),
                            ),
                            FlatButton(
                              child: Text(
                                '标',
                                style: TextStyle(
                                    color: _controller.value.bitrateIndex == 0
                                        ? Colors.yellow
                                        : Colors.green),
                              ),
                              onPressed: () {
                                _controller.setBitrateIndex(0);
                              },
                            ),
                            FlatButton(
                              child: Text(
                                '高',
                                style: TextStyle(
                                    color: _controller.value.bitrateIndex == 1
                                        ? Colors.yellow
                                        : Colors.green),
                              ),
                              onPressed: () {
                                _controller.setBitrateIndex(1);
                              },
                            ),
                            FlatButton(
                              child: Text(
                                '超',
                                style: TextStyle(
                                    color: _controller.value.bitrateIndex == 2
                                        ? Colors.yellow
                                        : Colors.green),
                              ),
                              onPressed: () {
                                _controller.setBitrateIndex(2);
                              },
                            ),
                          ],
                        ),
                        Row(
                          children: <Widget>[
                            Container(
                              margin: EdgeInsets.only(left: 15),
                              child: Text(
                                '普通点播 : ',
                                style: TextStyle(color: Colors.orange),
                              ),
                            ),
                            FlatButton(
                                onPressed: () {
                                  _controller = TencentPlayerController.network(
                                      spe1,
                                      playerConfig: PlayerConfig(
                                          startTime: _controller
                                              .value.position.inSeconds));
                                  _controller.initialize().then((_) {
                                    setState(() {});
                                  });
                                  //   _controller.addListener(listener);
                                },
                                child: Text(
                                  '标',
                                  style: TextStyle(
                                      color: _controller.dataSource == videoUrlB
                                          ? Colors.red
                                          : Colors.blue),
                                )),
                            FlatButton(
                                onPressed: () {
                                  _controller = TencentPlayerController.network(
                                      spe2,
                                      playerConfig: PlayerConfig(
                                          startTime: _controller
                                              .value.position.inSeconds));
                                  _controller.initialize().then((_) {
                                    setState(() {});
                                  });
                                  //    _controller.addListener(listener);
                                },
                                child: Text(
                                  '高',
                                  style: TextStyle(
                                      color: _controller.dataSource == videoUrlG
                                          ? Colors.red
                                          : Colors.blue),
                                )),
                            FlatButton(
                              onPressed: () {
                                _controller = TencentPlayerController.network(
                                    spe3,
                                    playerConfig: PlayerConfig(
                                        startTime: _controller
                                            .value.position.inSeconds));
                                _controller.initialize().then((_) {
                                  setState(() {});
                                });
                                //  _controller.addListener(listener);
                              },
                              child: Text(
                                '超',
                                style: TextStyle(
                                    color: _controller.dataSource == videoUrl
                                        ? Colors.red
                                        : Colors.blue),
                              ),
                            ),
                          ],
                        ),
                        Row(
                          children: <Widget>[
                            FlatButton(
                              onPressed: () {
                                _downloadController.dowload(
                                    "4564972819220421305",
                                    quanlity: 2);
                              },
                              child: Text(
                                'download1',
                                style: TextStyle(color: Colors.blue),
                              ),
                            ),
                            FlatButton(
                              onPressed: () {
                                _downloadController
                                    .pauseDownload("4564972819220421305");
                              },
                              child: Text(
                                'download1 - stop',
                                style: TextStyle(color: Colors.blue),
                              ),
                            ),
                          ],
                        ),
                        Row(
                          children: <Widget>[
                            FlatButton(
                              onPressed: () {
                                _downloadController.dowload(testDownload);
                              },
                              child: Text(
                                'download2',
                                style: TextStyle(color: Colors.blue),
                              ),
                            ),
                            FlatButton(
                              onPressed: () {
                                _downloadController.pauseDownload(testDownload);
                              },
                              child: Text(
                                'download2 - stop',
                                style: TextStyle(color: Colors.blue),
                              ),
                            ),
                          ],
                        ),
                        Text(_downloadController.value != null
                            ? _downloadController.value.toString()
                            : '')
                      ],
                    ),
                  )
                ],
              ),
            ),
            floatingActionButton: FloatingActionButton(
              onPressed: () {
                setState(() {
                  _controller.value.isPlaying
                      ? _controller.pause()
                      : _controller.play();
                });
              },
              child: Icon(
                _controller.value.isPlaying ? Icons.pause : Icons.play_arrow,
              ),
            ),
          );
        },
      ),
    );
  }
}
